tinygrad/tensor.py

```diff
@@ -314,22 +316,59 @@ def num_zeros(step, dim_sz): return 0 if step == 1 or (y := dim_sz % step) == 0
       final_slice = reduce(operator.add, (((0, sh), (0, 1)) for sh in new_shape), ())
       sliced_tensor = reshaped_tensor.shrink(final_slice)
     final_shape = []
+    sub = [0] * len(tensor_found)
     it_shape = iter(new_shape)
-    for i in orig_slices:
+    for n,i in enumerate(orig_slices):
       if isinstance(i, (int, slice)):
         dim_shape = next(it_shape)
         if isinstance(i, slice): final_shape.append(dim_shape)
+        else: # i is int
+          if tensor_found: # for tensor indexing
+            for n_ in range(len(tensor_found)):
+              if tensor_found[n_][0] > n: sub[n_] -= 1
       else: # i is None
         final_shape.append(1)
-    return sliced_tensor.reshape(tuple(final_shape))  # Reshape
+    ret = sliced_tensor.reshape(tuple(final_shape))  # Reshape
+    # Fancy/tensor indexing
+    if tensor_found:
+      for i,s in enumerate(sub): tensor_found[i] = (tensor_found[i][0]+s, tensor_found[i][1])
+      dim = [i[0] for i in tensor_found]
+      idx = [(i[1] < 0).where(i[1]+ret.shape[i[0]], i[1]) for i in tensor_found]
+      dim_cond = dim[0] != 0 and dim != list(range(dim[0], dim[-1]+1)) and len(dim) != 1
+      for n,d in enumerate(dim):
+        if n == 0:
+          new_ret = ret._gather(idx=idx[n], dim=d)
+        else:
+          new_idx = idx[n].reshape(*[1]*dim[0], *idx[n].shape, *[1]*(ret.ndim-dim[0]-n))
+          arange = Tensor.arange(ret.shape[d], dtype=dtypes.int32, requires_grad=False).reshape(*[1]*(idx[n].ndim+d-n), ret.shape[d], *[1]*(ret.ndim-d-1))
+          new_ret = ((new_idx == arange) * new_ret).sum(idx[n].ndim+d-n)
```

Comment:
You should be able to move the first iteration of the loop out of the loop at just start the loop at 1.